lib/ruby_lsp/requests/support/rubocop_runner.rb

```diff
@@ -138,3 +164,107 @@ def file_finished(_file, offenses)
     end
   end
 end
+
+# Processed Source patch so that we can pass the existing AST to RuboCop without having to re-parse files a second time
+module ProcessedSourcePatch
+  extend T::Sig
+
+  sig do
+    params(
+      source: String,
+      ruby_version: Float,
+      path: T.nilable(String),
+      parser_engine: Symbol,
+      prism_result: T.nilable(Prism::ParseResult),
+    ).void
+  end
+  def initialize(source, ruby_version, path = nil, parser_engine: :parser_whitequark, prism_result: nil)
+    @prism_result = prism_result
+
+    # Invoking super will end up invoking our patched version of tokenize, which avoids re-parsing the file
+    super(source, Prism::Translation::Parser::VERSION_3_3, path, parser_engine: parser_engine)
+  end
+
+  sig { params(parser: T.untyped).returns(T::Array[T.untyped]) }
+  def tokenize(parser)
+    begin
+      ast, comments, tokens = parser.tokenize(@buffer, parse_result: @prism_result)
+      ast ||= nil
+    rescue Parser::SyntaxError
+      comments = []
+      tokens = []
+    end
+
+    ast&.complete!
+    tokens.map! { |t| RuboCop::AST::Token.from_parser_token(t) }
+
+    [ast, comments, tokens]
+  end
+
+  RuboCop::AST::ProcessedSource.prepend(self)
+end
+
+module Prism
```

Comment:
This whole patch can probably just move into prism. We can have the parser accept an AST. I don't think we should maintain this patch inside ruby-lsp.